File: src/main/kotlin/com/huawei/cangjie/psi/CjTypeParameterListOwnerStub.java
package com.huawei.cangjie.psi;

import com.huawei.cangjie.psi.stubs.CangJieStubWithFqName;
import com.huawei.cangjie.psi.stubs.elements.CjStubElementTypes;
import com.intellij.lang.ASTNode;
import com.intellij.psi.stubs.IStubElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;



public abstract class CjTypeParameterListOwnerStub<T extends CangJieStubWithFqName<?>>
        extends CjNamedDeclarationStub<T> implements CjTypeParameterListOwner {
    public CjTypeParameterListOwnerStub(@NotNull T stub, @NotNull IStubElementType nodeType) {
        super(stub, nodeType);
    }

    public CjTypeParameterListOwnerStub(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    @Nullable
    public CjTypeParameterList getTypeParameterList() {
        return getStubOrPsiChild(CjStubElementTypes.TYPE_PARAMETER_LIST);
    }

    @Override
    @Nullable
    public CjTypeConstraintList getTypeConstraintList() {
        return getStubOrPsiChild(CjStubElementTypes.TYPE_CONSTRAINT_LIST);
    }

    @Override
    @NotNull
    public List<CjTypeConstraint> getTypeConstraints() {
        CjTypeConstraintList typeConstraintList = getTypeConstraintList();
        if (typeConstraintList == null) {
            return Collections.emptyList();
        }
        return typeConstraintList.getConstraints();
    }

    @Override
    @NotNull
    public List<CjTypeParameter> getTypeParameters() {
        CjTypeParameterList list = getTypeParameterList();
        if (list == null) return Collections.emptyList();

        return list.getParameters();
    }
}

File: src/main/kotlin/com/huawei/cangjie/psi/CjTypeParameterListOwnerStub.kt
package com.huawei.cangjie.psi

import com.huawei.cangjie.psi.stubs.CangJieStubWithFqName
import com.huawei.cangjie.psi.stubs.elements.CjStubElementTypes
import com.intellij.lang.ASTNode
import com.intellij.psi.stubs.IStubElementType

abstract class CjTypeParameterListOwnerStub<T : CangJieStubWithFqName<*>>
    : CjNamedDeclarationStub<T>, CjTypeParameterListOwner {
    constructor(stub: T, nodeType: IStubElementType<*, *>) : super(stub, nodeType)

    constructor(node: ASTNode) : super(node)


    override val typeParameterList: CjTypeParameterList? get() = getStubOrPsiChild(CjStubElementTypes.TYPE_PARAMETER_LIST)
    override val typeConstraintList: CjTypeConstraintList? get() = getStubOrPsiChild(CjStubElementTypes.TYPE_CONSTRAINT_LIST)
    override val typeConstraints: List<CjTypeConstraint>
        get() {
            val typeConstraintList = typeConstraintList ?: return emptyList()
            return typeConstraintList.constraints
        }
    override val typeParameters: List<CjTypeParameter>
        get() {
            val list = typeParameterList ?: return emptyList()

            return list.parameters
        }

}
